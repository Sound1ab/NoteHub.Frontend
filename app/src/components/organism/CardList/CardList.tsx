import React from 'react'
import { Code } from 'react-content-loader'
import { useStore } from '../../../hooks'
import { useListFiles } from '../../../hooks/file/useListFiles'
import { activeNote } from '../../../store'
import { styled } from '../../../theme'
import { Container } from '../../atoms'
import { Card, CardHeader } from '../../molecules'

const Style = styled.div`
  position: relative;
  flex: 0 0 auto;
  width: ${({ theme }) => theme.spacing.xxxl}
  height: 100%;
  background-color: ${({ theme }) => theme.colors.background.secondary};
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;

  .CardList-wrapper {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
  }

  .CardList-card-wrapper {
    border-bottom: 1px solid ${({ theme }) => theme.colors.text.secondary};
  }
  
  .CardList-loader {
    margin-left: ${({ theme }) => theme.spacing.s};
    margin-top: ${({ theme }) => theme.spacing.s};
  }
`

export function CardList() {
  const [state, dispatch] = useStore()
  const { files, loading } = useListFiles(
    state.user.username,
    state.notebook.activeNotebook
  )

  function handleCardClick(note: string | null) {
    if (dispatch) dispatch(activeNote(note))
  }

  return (
    <Style>
      <CardHeader title={state.notebook.activeNotebook} />
      <Container className="CardList-wrapper">
        {loading ? (
          <>
            <Code className="CardList-loader" />
            <Code className="CardList-loader" />
          </>
        ) : (
          files
            .sort((noteA, noteB) => {
              return noteA.filename.localeCompare(noteB.filename)
            })
            .map(note => {
              return (
                <span
                  key={note.sha}
                  className="CardList-card-wrapper"
                  onClick={handleCardClick.bind(null, note.filename)}
                >
                  <Card
                    key={note.sha}
                    id={note.sha}
                    title={note.filename}
                    excerpt={note.excerpt || ''}
                    githubLink={note._links.html}
                    createdAt={''}
                    isSelected={
                      !!state.notebook.activeNote &&
                      state.notebook.activeNote === note.filename
                    }
                  />
                </span>
              )
            })
        )}
      </Container>
    </Style>
  )
}
