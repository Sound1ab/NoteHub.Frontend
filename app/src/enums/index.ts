export enum EXEC_COMMAND {
  UNDERLINE = 'underline',
}

export enum COLOR {
  DARK = 'dark',
  LIGHT = 'light',
  ACCENT = 'accent',
  MEDIUM = 'medium',
}

export enum LOCAL_STORAGE {
  KEY = 'noted::access_token',
}

export enum GRAPHQL {
  DEV_GRAPHQL = process.env.REACT_APP_GRAPHQL_DEV! as any,
  PROD_GRAPHQL = process.env.NODE_ENV! as any,
}

export enum APOLLO_ERRORS {
  UNAUTHENTICATED = 'UNAUTHENTICATED',
}
