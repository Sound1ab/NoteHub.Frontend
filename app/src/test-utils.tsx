import * as React from 'react'
import { render } from 'react-testing-library'
import { ThemeProvider } from './components/utility'

const customRender = (node: any, ...options: any[]) =>
  render(<ThemeProvider>{node}</ThemeProvider>, ...options)

// re-export everything
export * from 'react-testing-library'

// override render method
export { customRender as render }
