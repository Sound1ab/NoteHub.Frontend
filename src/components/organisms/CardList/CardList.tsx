import React from 'react'

import {
  useListFiles,
  useReadCurrentFileName,
  useReadIsNewFileOpen,
} from '../../../hooks'
import { styled } from '../../../theme'
import { Card, CardSkeleton, FileInput } from '../../molecules'

const Style = styled.div`
  grid-area: cardlist;
  position: relative;
  flex: 0 0 auto;
  height: 100%;
  background-color: ${({ theme }) => theme.colors.background.primary};
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
  border-right: 1px solid ${({ theme }) => theme.colors.border};
  overflow: auto;
  resize: horizontal;
  min-width: ${({ theme }) => theme.spacing.xxl};
  max-width: 50vw;
`

export function CardList() {
  const { currentFileName, client } = useReadCurrentFileName()
  const { files, loading } = useListFiles()
  const { isNewFileOpen } = useReadIsNewFileOpen()

  function handleCardClick(filename: string) {
    client.writeData({ data: { currentFileName: filename } })
  }

  return (
    <Style>
      {isNewFileOpen && <Card renderInput={<FileInput />} />}
      {loading ? (
        <CardSkeleton />
      ) : (
        files
          .sort((fileA, fileB) => {
            return fileA.filename.localeCompare(fileB.filename)
          })
          .map(file => {
            return (
              <Card
                key={`${file.sha}-${file.filename}`}
                onClick={() => handleCardClick(file.filename)}
                heading={file.filename.replace(/.md/, '')}
                isActive={currentFileName === file.filename}
                isDisabled={file.sha === 'optimistic'}
              />
            )
          })
      )}
    </Style>
  )
}
