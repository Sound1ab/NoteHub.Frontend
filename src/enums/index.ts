export enum TOGGLES {
  ROCK = 'rock',
  FOLK = 'folk',
  JAZZ = 'jazz',
  ALL = 'all',
}
