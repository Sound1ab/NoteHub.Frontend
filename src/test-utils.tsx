import '@testing-library/jest-dom/extend-expect'

import { EventType } from '@testing-library/dom/types/events'
import { act, fireEvent, render } from '@testing-library/react'
import React, { ReactNode } from 'react'

import {
  IconProvider,
  MockProvider,
  ThemeProvider,
} from './components/providers'
import { FileTreeProvider } from './components/templates/Dashboard/Sidebar/FileTree/FileTreeProvider'

export type FireObject = {
  [K in EventType]: (
    element: Document | Element | Window,
    options?: Record<string, unknown>
  ) => Promise<boolean>
}
global.matchMedia = () =>
  ({
    matches: false,
  } as MediaQueryList)

const wait = (): Promise<void> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve()
    }, 1)
  })
}

const Context = ({ node }: { node: ReactNode }) => (
  <MockProvider>
    <ThemeProvider>
      {() => (
        <IconProvider>
          <FileTreeProvider>{node}</FileTreeProvider>
        </IconProvider>
      )}
    </ThemeProvider>
  </MockProvider>
)

const customRender = async (
  node: ReactNode,
  { waitForLoad = true, ...options } = {}
) => {
  const { rerender, ...rest } = render(<Context node={node} />, options)

  if (waitForLoad) {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    await act(wait)
  }

  return {
    ...rest,
    rerender: async (node: ReactNode) => {
      rerender(<Context node={node} />)

      // eslint-disable-next-line @typescript-eslint/no-empty-function
      await act(wait)
    },
  }
}

const customFireEvent = Object.entries(fireEvent).reduce(
  (
    acc: Record<
      string,
      (
        element: Document | Element | Window,
        options: Record<string, string>
      ) => Promise<unknown>
    >,
    keyValue
  ) => {
    const [key, value] = keyValue
    acc[key] = async (
      element: Document | Element | Window,
      options: Record<string, unknown>
    ) => {
      if (typeof value === 'function') {
        value(element, options)
      }
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      await act(wait)
    }
    return acc
  },
  {}
)

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

export { customFireEvent as fireEvent }
