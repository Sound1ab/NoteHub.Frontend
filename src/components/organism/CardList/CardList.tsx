import React from 'react'
import { TOGGLES } from '../../../enums'
import { IEvent } from '../../../interfaces'
import { styled } from '../../../theme'
import { Card } from '../../molecules'

const Style = styled.div`
  position: relative;

  > div {
    margin-top: ${({ theme }) => theme.spacing.xs};
  }
`

interface ICardList {
  events: IEvent[]
  activeToggle: TOGGLES
}

export function CardList({ events, activeToggle }: ICardList) {
  return (
    <Style>
      {events
        .filter(event =>
          activeToggle === TOGGLES.ALL ? event : event.genre === activeToggle
        )
        .sort(
          (eventA, eventB) => eventA.date.dayOfMonth - eventB.date.dayOfMonth
        )
        .map(event => (
          <Card
            key={`${event.date.dayOfMonth}-${event.location}-${activeToggle}`}
            dayOfMonth={event.date.dayOfMonth}
            dayOfWeek={event.date.dayOfWeek}
            location={event.location}
            month={event.date.month}
          />
        ))}
    </Style>
  )
}
