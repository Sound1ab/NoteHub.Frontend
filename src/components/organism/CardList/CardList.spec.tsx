import 'jest-dom/extend-expect'
import React from 'react'
import { TOGGLES } from '../../../enums'
import { cleanup, render } from '../../../test-utils'
import { CardList } from './CardList'

const MOCK_EVENTS = [
  {
    date: {
      dayOfMonth: 13,
      dayOfWeek: 'Thursday',
      month: 'Oct',
    },
    genre: 'rock',
    location: 'Whitechapel',
  },
  {
    date: {
      dayOfMonth: 13,
      dayOfWeek: 'Thursday',
      month: 'Oct',
    },
    genre: 'jazz',
    location: 'Hoxton',
  },
]
const MOCK_TOGGLE = TOGGLES.JAZZ

afterEach(cleanup)

describe('CardList.tsx', () => {
  test('CardList matches snapshot', () => {
    const { asFragment } = render(
      <CardList events={MOCK_EVENTS} activeToggle={MOCK_TOGGLE} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('CardList filters cards based on toggle', async () => {
    const { getAllByTestId } = render(
      <CardList events={MOCK_EVENTS} activeToggle={MOCK_TOGGLE} />
    )

    const cards = getAllByTestId('Card')
    expect(cards.length).toBe(1)
    expect(cards[0]).toHaveTextContent('Hoxton')
    expect(cards[0]).not.toHaveTextContent('Whitechapel')
  })

  test('Cardlist shows all on all toggle', () => {
    const MOCK_ALL_TOGGLE = TOGGLES.ALL
    const { getAllByTestId } = render(
      <CardList events={MOCK_EVENTS} activeToggle={MOCK_ALL_TOGGLE} />
    )

    const cards = getAllByTestId('Card')
    expect(cards.length).toBe(2)
    expect(cards[0]).toHaveTextContent('Whitechapel')
    expect(cards[1]).toHaveTextContent('Hoxton')
  })
})
