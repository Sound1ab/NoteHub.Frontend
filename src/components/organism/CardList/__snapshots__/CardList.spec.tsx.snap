// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList.tsx CardList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bZQynM bRkGRU"
  >
    <div
      class="sc-bxivhb FDbHW"
      data-testid="Card"
    >
      <div
        class="card-cell card-date"
      >
        <h2
          class="sc-gzVnrw fDlMua"
          type="h2"
        >
          13
        </h2>
        <h6
          class="sc-htoDjs eWaNkr"
          type="h6"
        >
          Oct
        </h6>
      </div>
      <div
        class="card-cell card-location"
      >
        <h6
          class="sc-dnqmqq fWMmgG"
          type="h6"
        >
          <span
            class="card-highlight"
          >
            Thursday
          </span>
        </h6>
        <h2
          class="sc-iwsKbI caDdqp"
          type="h2"
        >
          Hoxton
        </h2>
      </div>
      <span
        class="card-chevron"
      />
    </div>
  </div>
</DocumentFragment>
`;
